server/base/src/main/java/org/apache/accumulo/server/util/Admin.java

```diff
@@ -563,18 +554,89 @@ private static void stopServer(final ClientContext context, final boolean tablet
         client -> client.shutdown(TraceUtil.traceInfo(), context.rpcCreds(), tabletServersToo));
   }
 
-  // Visible for tests
-  public static void signalGracefulShutdown(final ClientContext context, String address) {
+  private static void stopServers(final ServerContext context, List<String> servers,
+      final boolean force)
+      throws AccumuloException, AccumuloSecurityException, InterruptedException, KeeperException {
+    List<String> hostOnly = new ArrayList<>();
+    Set<HostAndPort> hostAndPort = new TreeSet<>();
 
-    Objects.requireNonNull(address, "address not set");
-    final HostAndPort hp = HostAndPort.fromString(address);
+    for (var server : servers) {
+      if (server.contains(":")) {
+        hostAndPort.add(HostAndPort.fromString(server));
+      } else {
+        hostOnly.add(server);
+      }
+    }
+
+    if (!hostOnly.isEmpty()) {
+      // TODO need to see how easy it is to use serverStatus to support current functionality of
+      // specifying a host. Like what does the whole command to stop all tservers on a host look
+      // like using admin serviceStatus and admin stop.
+      // TODO start deprecation warning in 4.0 instead of 2.1.x?
```

Comment:
maybe in 4.0 the port is optional, instead of deprecated. Understood that you are trying to preserve behavior in the patch release when port is not specified. In 4.0, we could just change the behavior such that if there is no port, everything on the node gets shut down.

Reply:
I pulled the TODO and warnings out for 2.1.  We definitely  need to do something in 4.0.  The way the current code uses the tserver port config is a sharp edge, it may not actually stop everything on the host.  I tried starting three tservers and then ran stop w/ just the host and it only stopped one based on my port config.  I did not have a port range, but I did have port search enabled.  I will open a follow on issue for 4.0 to change the behavior for  host only.